### ISSUE TITLE
Weapon/ammo count doesn't match between map changes

### ISSUE THREAD

**Issue by TetraTheta** [ampreeT/SourceCoop#59]
Server: Windows 11

SourceCoop doesn't give weapon and its ammo count back to player between map trasition. Players get weapon that he didn't get in previous map, and other player lost his weapon that he had in previous map.

My guess is, SC just distributes weapons to player with already-made list. So, if player accuired weapon in previous map but that weapon doesn't in the list, he will just lost his weapon.

PS. `sourcecoop_ft KEEP_EQUIPMENT 1` was enabled.

**Comment by Alienmario** [comment]
This is intentional. Item persistence might be added in the future.

**Comment by raicovx** [comment]
What would be the ideal config for item persistence?

Currently I'm working on creating something rudimentary for a playthrough with friends soon, but i'd be happy to contribute. Admittedly its my first attempt at sourcemod, though I've dabbled in c++ before and work with c# full-time. 

So far I've started implementing a basic KeyValue system for it, saving data per level:
https://gist.github.com/raicovx/7dfbfb16d88c386156d0fe5f2826e873

**Comment by TetraTheta** [comment]
> What would be the ideal config for item persistence?
Havn't looked in the link(outside of my home), but I think saving two data - What weapon the player have, What amount of ammunition the player have - is good way to making it persistent.

In basic, there is already list of what kind of weapon exists in Black Mesa, and by looking into weapon script, we can find out what ammunition type it uses.

But, each map already provide some 'default weapon / ammunition', I think those should be stripped out for consistency. This may result some visual problem like 'weapons shown in hand, disappears, and re-appear', but I'm not sure if it would really happen.

Thanks for the contribution!

**Comment by raicovx** [comment]
All good, so far i'm just kind of checkpointing what weapons, hp, armor and ammo each player has in a KeyValue file per map change. I plan to add a condition that should override the default map equipment from what I can see, but I'm not that far along yet. I'll update here once I have something functioning :)

**Comment by Alienmario** [comment]
You don't want to store it in keyvalues. That's an unnecessarily heavy system. Keep it in memory.

I think you may be able to reuse the enum struct CCoopEquipment in checkpoint.inc. Then you'd have a [StringMap](https://sm.alliedmods.net/new-api/adt_trie/StringMap) instance that keeps the steam id as key and the enum struct as value (enum struct are saved with StringMap.SetArray).

For equipping, modify the function CCoopSpawn::SpawnPlayerEquipment. You will have to account for midgame joiners and give them the default equipment. That will create imbalances but I guess that's expected.

**Comment by raicovx** [comment]
That definitely sounds like a better approach. For my use case I also wanted to be able to persist across multiple play sessions (I stop the server in between). So I might look into exporting / importing the data with a save / load command. Unless there is a better way I don't know about of course.

Thanks for the help :)

**Comment by raicovx** [comment]
I've got it working here:
https://github.com/raicovx/SourceCoop/tree/equipment_manager

I don't do anything special to handle map changes via vote etc, so I'm sure there are other considerations to build upon this. Also, I had an issue where the HEV suit just wouldn't be given to the player (even though it appears to be handled separately from the checkpoint.inc ParseEquipment). So I just manually give the player a hev suit if its not one of the starting levels. I imagine I'm missing something there so i'd appreciate any guidance to correct that.

**Comment by Alienmario** [comment]
So the way it works currently on *most* maps, is that equipment is scanned from entities in the map (as specified by edt scripts). This way you equip the same items as when launching a random map in singleplayer. Those entities are then removed from the map and added as weapons in the default equipment. This also includes suit, ammo, suit batteries,.. This method has worked so far, the game equips all items with a single function. However I could also see this being better organized if need be.

tldr: Easiest solution is to save item_suit as a weapon.
